File: src/audio/crossfade/wsolaPlanner.js
/*
 * Copyright (C) 2026 SFG545
 *
 * This file is part of Orchard.
 *
 * Orchard is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option) any
 * later version.
 *
 * Orchard is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with Orchard. If not, see <https://www.gnu.org/licenses/>.
 */

// This module is deliberately an adapter. PairTransitionPlanner owns every
// musical choice; this file only exposes its exact native-render fields in the
// shape the existing preparation/playback path understands.
import { normalizeTrackAnalysis } from '../../../shared/trackAnalysis.js';
import { planPairTransition } from './pairTransitionPlanner.js';
export { alignTempoOctave } from './transitionPolicy.js';

const SLICE_PADDING_SECONDS = 1.5;
const HANDOFF_FRACTION = 0.5;
const BED_POSITION = 0.5;

function finite(value) {
  const number = Number(value);
  return Number.isFinite(number) ? number : 0;
}

function canonicalAnalysis(value = {}, duration = 0) {
  if (
    value?.timing &&
    value?.audibleRange &&
    Array.isArray(value?.frames) &&
    Array.isArray(value?.boundaries)
  ) return value;
  return normalizeTrackAnalysis({
    ...value,
    duration: Math.max(finite(value.duration), finite(duration))
  });
}

function sliceFor(start, end, duration) {
  const sliceStart = Math.max(0, start - SLICE_PADDING_SECONDS);
  const sliceEnd = Math.min(duration || Infinity, end + SLICE_PADDING_SECONDS);
  return {
    start: sliceStart,
    end: sliceEnd,
    anchor: start - sliceStart
  };
}

function refusal(pairPlan) {
  return {
    ok: false,
    reason: pairPlan.fallbackReason || pairPlan.transitionClass,
    fallback: pairPlan.fallback,
    pairPlan
  };
}

export function planWsolaTransition({
  analysis = {},
  nextAnalysis = {},
  duration = 0,
  nextDuration = 0
} = {}) {
  const pairPlan = planPairTransition({ analysis, nextAnalysis, duration, nextDuration });
  if (pairPlan.renderMode !== 'native') return refusal(pairPlan);

  const outgoing = canonicalAnalysis(analysis, duration);
  const incoming = canonicalAnalysis(nextAnalysis, nextDuration);
  const outgoingLength = Math.max(finite(duration), finite(outgoing.duration));
  const incomingLength = Math.max(finite(nextDuration), finite(incoming.duration));
  const outgoingGrid = {
    bpm: finite(outgoing.timing?.bpm),
    beats: [...(outgoing.timing?.beats || [])],
    downbeats: [...(outgoing.timing?.downbeats || [])]
  };
  const incomingGrid = {
    bpm: finite(incoming.timing?.bpm),
    beats: [...(incoming.timing?.beats || [])],
    downbeats: [...(incoming.timing?.downbeats || [])]
  };
  const selectedSources = pairPlan.diagnostics?.selected?.sources || [];
  const selectedGates = pairPlan.diagnostics?.selected?.gates || [];

  return {
    ok: true,
    pairPlan,
    fallback: pairPlan.fallback,
    outgoingGrid,
    incomingGrid,
    tier: 'beatmatched',
    beatConfidence: Math.min(
      finite(outgoing.timing?.beatConfidence),
      finite(incoming.timing?.beatConfidence)
    ),
    mixOutType: selectedSources[0] || '',
    vocalClash: selectedGates.includes('vocal-collision'),
    transitionStart: pairPlan.outgoing.start,
    transitionEnd: pairPlan.outgoing.end,
    overlapSeconds: pairPlan.durationSeconds,
    beats: pairPlan.beats,
    fadeBeats: pairPlan.beats,
    handoffFraction: HANDOFF_FRACTION,
    bedPosition: BED_POSITION,
    bassSwapFraction: pairPlan.strategy === 'bass_swap' ? 0.5 : 0.7,
    filterSweep: pairPlan.rendering.filters ? 1 : 0,
    outgoingBpm: outgoingGrid.bpm,
    incomingBpm: incomingGrid.bpm,
    targetBpm: pairPlan.targetBpm,
    outgoingTempoRatio: pairPlan.outgoing.tempoRatio,
    incomingTempoRatio: pairPlan.incoming.tempoRatio,
    stretchRatio: pairPlan.outgoing.tempoRatio,
    incomingCueTime: pairPlan.incoming.start,
    incomingDropTime: pairPlan.incoming.handoff,
    incomingResumeTime: pairPlan.incoming.resume,
    strategy: pairPlan.strategy,
    outgoingSlice: sliceFor(
      pairPlan.outgoing.start,
      pairPlan.outgoing.end,
      outgoingLength
    ),
    incomingSlice: sliceFor(
      pairPlan.incoming.start,
      pairPlan.incoming.resume,
      incomingLength
    )
  };
}

File: mobile/android/app/src/main/java/dev/sfg/orchard/mobile/playback/smart/WsolaPlanner.kt
/*
 * Copyright (C) 2026 SFG545
 *
 * This file is part of Orchard.
 *
 * Orchard is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option) any
 * later version.
 *
 * Orchard is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with Orchard. If not, see <https://www.gnu.org/licenses/>.
 */

package dev.sfg.orchard.mobile.playback.smart

const val HANDOFF_FRACTION = 0.5
const val BED_POSITION = 0.5
const val FILTER_SWEEP = 1.0

sealed interface WsolaPlanResult {
    data class Refused(val reason: String) : WsolaPlanResult

    /** All times are seconds on each track's own media timeline. */
    data class Planned(
        val tier: TransitionTier,
        val beatConfidence: Double,
        val mixOutType: String,
        /** True when the fade was shortened all the way to the floor and both sides still sing. */
        val vocalClash: Boolean,
        val transitionStart: Double,
        val transitionEnd: Double,
        val overlapSeconds: Double,
        val beats: Int,
        /**
         * What the fade actually spends, after the intro has had its say. The pre-clamp target is
         * not reported: nothing downstream can act on beats the incoming track had no room for.
         */
        val fadeBeats: Int,
        val handoffFraction: Double,
        val bedPosition: Double,
        val bassSwapFraction: Double,
        val filterSweep: Double,
        val outgoingBpm: Double,
        val incomingBpm: Double,
        /** Source seconds consumed by the outgoing deck per rendered output second. */
        val stretchRatio: Double,
        val incomingCueTime: Double,
        val incomingDropTime: Double,
        /** Where the blend finishes on the incoming timeline, after the arrangement has arrived. */
        val incomingHandoffTime: Double,
        val incomingResumeTime: Double,
        val choreography: TransitionChoreography? = null,
        val targetBpm: Double = incomingBpm,
        val outgoingTempoRatio: Double = stretchRatio,
        val incomingTempoRatio: Double = 1.0,
        val strategy: String = "beatmatched_crossfade",

    ) : WsolaPlanResult
}


/** The desktop native adapter, without mobile offsets or an additional policy gate. */
fun planWsolaTransition(
    analysis: TrackAnalysis, nextAnalysis: TrackAnalysis, duration: Double = 0.0, nextDuration: Double = 0.0,
): WsolaPlanResult {
    val result = DesktopTransitionPlanner.invoke("native", plannerInput(analysis, nextAnalysis, duration, nextDuration))
    if (!result.getBoolean("ok")) return WsolaPlanResult.Refused(result.getString("reason"))
    return WsolaPlanResult.Planned(
        tier = TransitionTier.BEATMATCHED, beatConfidence = result.getDouble("beatConfidence"),
        mixOutType = result.getString("mixOutType"), vocalClash = result.getBoolean("vocalClash"),
        transitionStart = result.getDouble("transitionStart"), transitionEnd = result.getDouble("transitionEnd"),
        overlapSeconds = result.getDouble("overlapSeconds"), beats = result.getInt("beats"),
        fadeBeats = result.getInt("fadeBeats"), handoffFraction = result.getDouble("handoffFraction"),
        bedPosition = result.getDouble("bedPosition"), bassSwapFraction = result.getDouble("bassSwapFraction"),
        filterSweep = result.getDouble("filterSweep"), outgoingBpm = result.getDouble("outgoingBpm"),
        incomingBpm = result.getDouble("incomingBpm"), stretchRatio = result.getDouble("stretchRatio"),
        incomingCueTime = result.getDouble("incomingCueTime"), incomingDropTime = result.getDouble("incomingDropTime"),
        incomingHandoffTime = result.getDouble("incomingResumeTime"), incomingResumeTime = result.getDouble("incomingResumeTime"),
        choreography = result.getJSONObject("choreography").choreography(), targetBpm = result.getDouble("targetBpm"),
        outgoingTempoRatio = result.getDouble("outgoingTempoRatio"), incomingTempoRatio = result.getDouble("incomingTempoRatio"),
        strategy = result.getString("strategy"),
    )
}

fun incomingAudibleStart(analysis: TrackAnalysis): Double = audibleStartOf(analysis)
